crates/publish/src/oci/client.rs

```diff
@@ -0,0 +1,261 @@
+            self.oci
+                .pull_blob(&reference, &layer.digest, &mut bytes)
+                .await?;
+
+            match layer.media_type.as_str() {
+                WASM_LAYER_MEDIA_TYPE => self.cache.write_wasm(&bytes, &layer.digest).await?,
+                _ => self.cache.write_data(&bytes, &layer.digest).await?,
+            }
+        }
+
+        tracing::info!("Pulled {}@{}", reference, digest);
+
+        Ok(())
+    }
+
+    /// Create a new wasm layer based on a file.
+    pub async fn wasm_layer(file: &Path) -> Result<ImageLayer> {
+        tracing::log::trace!("Reading wasm module from {:?}", file);
+        Ok(ImageLayer::new(
+            fs::read(file).await.context("cannot read wasm module")?,
+            WASM_LAYER_MEDIA_TYPE.to_string(),
+            None,
+        ))
+    }
+
+    /// Create a new data layer based on a file.
+    pub async fn data_layer(file: &Path) -> Result<ImageLayer> {
+        tracing::log::trace!("Reading data file from {:?}", file);
+        Ok(ImageLayer::new(
+            fs::read(&file).await?,
+            DATA_MEDIATYPE.to_string(),
+            None,
+        ))
+    }
+
+    /// Construct the registry authentication based on the reference.
+    pub fn auth(reference: &Reference) -> Result<RegistryAuth> {
+        let server = reference
+            .resolve_registry()
+            .strip_suffix('/')
+            .unwrap_or_else(|| reference.resolve_registry());
+
+        let creds = docker_credential::get_credential(server);
+        match creds {
+            Err(CredentialRetrievalError::ConfigNotFound) => Ok(RegistryAuth::Anonymous),
+            Err(CredentialRetrievalError::NoCredentialConfigured) => Ok(RegistryAuth::Anonymous),
+            Err(CredentialRetrievalError::ConfigReadError) => Ok(RegistryAuth::Anonymous),
+            Err(e) => bail!("Error handling docker configuration file: {}", e),
+
+            Ok(DockerCredential::UsernamePassword(username, password)) => {
+                tracing::trace!("Found docker credentials");
+                Ok(RegistryAuth::Basic(username, password))
+            }
+            Ok(DockerCredential::IdentityToken(_)) => {
+                println!("Cannot use contents of docker config, identity token not supported. Using anonymous auth");
+                Ok(RegistryAuth::Anonymous)
+            }
+        }
+    }
+
+    /// Build the OCI client configuration given the insecure option.
+    pub fn build_config(insecure: bool) -> oci_distribution::client::ClientConfig {
+        let protocol = if insecure {
+            oci_distribution::client::ClientProtocol::Http
+        } else {
+            oci_distribution::client::ClientProtocol::Https
+        };
+
+        oci_distribution::client::ClientConfig {
+            protocol,
+            ..Default::default()
+        }
+    }
+
+    /// Read the contents of a file and return it as a byte array.
+    pub async fn data(file: &Path) -> Result<Vec<u8>> {
```

Comment:
Not sure quite why this exists - is it anything more than a renaming of `fs::read`?  If it is needed, suggest `read` or `read_file` so the name isn't entirely novel, and maybe a comment to explain why the wrapper is needed?

Reply:
I changed all call sites of the function to use `tokio` directly in a previous commit, but forgot to remove the function.
Good catch.